File: src/main/java/cyclicsort/FirstMissingPositive.java
package cyclicsort;

import util.Util;

public class FirstMissingPositive {
    public static int sol(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            // i = nums[i] - 1
            while (nums[i] > 0 && nums[i] < nums.length && nums[i] != nums[nums[i] - 1]) {
                Util.swap(nums, i, nums[i] - 1);
            }
        }
        for (int i = 0; i < nums.length; i++) {
            if (i != nums[i] - 1) {
                return i + 1;
            }
        }

        // 1, 2, 3, 4 -> return length + 1 == 5
        return nums.length + 1;
    }
}

File: src/main/java/cyclicsort/FirstMissingPositive.py
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # index = 0, 1, 2, len(nums) - 1
        # value = 1, 2, 3, len(nums)
        # index == value - 1

        # Swap the current value to where it should be

        for i in range(len(nums)):
            current = nums[i]
            index = nums[i] - 1
            while 0 < current <= len(nums) and current != nums[index]:
                i, index = index, i


        for i in range(len(nums)):
            if i != nums[i] - 1:
                return i + 1
        return len(nums) + 1

